File: patterns/javascript/Patterns.js
const inverted = (n) => {
    for (let i = n; i >= 1; i--) {
        console.log(" ".repeat(n - i) + "*".repeat(2 * i - 1));
    }
};

const diamond = (n) => {
    pyramid(n);
    for (let i = n - 1; i >= 1; i--) {
        console.log(" ".repeat(n - i) + "*".repeat(2 * i - 1));
    }
};

const hollowPyramid = (n) => {
    for (let i = 1; i <= n; i++) {
        if (i === 1) {
            console.log(" ".repeat(n - 1) + "*");
        } else if (i === n) {
            console.log("*".repeat(2 * n - 1));
        } else {
            console.log(" ".repeat(n - i) + "*" + " ".repeat(2 * i - 3) + "*");
        }
    }
};

const pascals = (n) => {
    let row = [1];
    for (let i = 0; i < n; i++) {
        const rowStr = row.join(" ");
        const padding = " ".repeat(Math.floor((2 * n - rowStr.length) / 2));
        console.log(padding + rowStr);

        const nextRow = [1];
        for (let j = 0; j < row.length - 1; j++) {
            nextRow.push(row[j] + row[j + 1]);
        }
        nextRow.push(1);
        row = nextRow;
    }
};

const hourglass = (n) => {
    inverted(n);
    for (let i = 2; i <= n; i++) {
        console.log(" ".repeat(n - i) + "*".repeat(2 * i - 1));
    }
};

const zigzag = (n) => {
    const height = 4;
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < height; j++) {
            if (j === 0 || j === height - 1) {
                console.log("*".repeat(n));
            } else {
                console.log(" ".repeat(n - i - 1) + "*");
            }
        }
    }
};


const main = () => {
    const args = process.argv.slice(2);
    const pattern = args[0] || "pyramid";
    const n = parseInt(args[1], 10) || 5;

    const funcs = {
        pyramid,
        inverted,
        diamond,
        hollow_pyramid: hollowPyramid,
        pascals,
        hourglass,
        zigzag,
    };

    if (funcs[pattern]) {
        funcs[pattern](n);
    } else {
        console.log("Unknown pattern. Choose from:", Object.keys(funcs).join(", "));
    }
};

main();

File: patterns/csharp/Patterns.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Patterns
{
    private static void PrintChars(char c, int n)
    {
        if (n > 0)
        {
            Console.Write(new string(c, n));
        }
    }

    public static void Pyramid(int n)
    {
        for (int i = 1; i <= n; i++)
        {
            PrintChars(' ', n - i);
            PrintChars('*', 2 * i - 1);
            Console.WriteLine();
        }
    }

    public static void Inverted(int n)
    {
        for (int i = n; i >= 1; i--)
        {
            PrintChars(' ', n - i);
            PrintChars('*', 2 * i - 1);
            Console.WriteLine();
        }
    }

    public static void Diamond(int n)
    {
        Pyramid(n);
        for (int i = n - 1; i >= 1; i--)
        {
            PrintChars(' ', n - i);
            PrintChars('*', 2 * i - 1);
            Console.WriteLine();
        }
    }
    
    public static void HollowPyramid(int n)
    {
        for (int i = 1; i <= n; i++)
        {
            PrintChars(' ', n - i);
            if (i == 1 || i == n)
            {
                PrintChars('*', 2 * i - 1);
            }
            else
            {
                Console.Write("*");
                PrintChars(' ', 2 * i - 3);
                Console.Write("*");
            }
            Console.WriteLine();
        }
    }

    public static void Pascals(int n)
    {
        List<long> row = new List<long> { 1 };
        for (int i = 0; i < n; i++)
        {
            string rowStr = string.Join(" ", row);
            int padding = (2 * n - rowStr.Length) / 2;
            PrintChars(' ', padding);
            Console.WriteLine(rowStr);

            List<long> nextRow = new List<long> { 1 };
            for (int j = 0; j < row.Count - 1; j++)
            {
                nextRow.Add(row[j] + row[j + 1]);
            }
            nextRow.Add(1);
            row = nextRow;
        }
    }

    public static void Hourglass(int n)
    {
        Inverted(n);
        for (int i = 2; i <= n; i++)
        {
            PrintChars(' ', n - i);
            PrintChars('*', 2 * i - 1);
            Console.WriteLine();
        }
    }
    
    public static void Zigzag(int n)
    {
        int height = 4;
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < height; j++)
            {
                if (j == 0 || j == height - 1)
                {
                    PrintChars('*', n);
                    Console.WriteLine();
                }
                else
                {
                    PrintChars(' ', n - i - 1);
                    Console.WriteLine("*");
                }
            }
        }
    }

    public static void Main(string[] args)
    {
        string pattern = args.Length > 0 ? args[0] : "pyramid";
        int n = 5;
        if (args.Length > 1 && int.TryParse(args[1], out int parsedN))
        {
            n = parsedN;
        }

        var funcs = new Dictionary<string, Action<int>>
        {
            ["pyramid"] = Pyramid,
            ["inverted"] = Inverted,
            ["diamond"] = Diamond,
            ["hollow_pyramid"] = HollowPyramid,
            ["pascals"] = Pascals,
            ["hourglass"] = Hourglass,
            ["zigzag"] = Zigzag,
        };

        if (funcs.ContainsKey(pattern))
        {
            funcs[pattern](n);
        }
        else
        {
            Console.WriteLine($"Unknown pattern. Choose from: {string.Join(", ", funcs.Keys)}");
        }
    }
}